pymc/aesaraf.py

```diff
@@ -1015,3 +1019,144 @@ def compile_pymc(
+    that might lead to different variable values from those that could have been seen during
+    inference, as such, **any variable that is passed in the ``givens_dict`` will be considered
+    volatile**.
+
+    Parameters
+    ----------
+    outputs : List[aesara.graph.basic.Variable]
+        The list of variables that will be returned by the compiled function
+    vars_in_trace : List[aesara.graph.basic.Variable]
+        The list of variables that are assumed to have values stored in the trace
+    givens_dict : Optional[Dict[aesara.graph.basic.Variable, Any]]
+        A dictionary that maps tensor variables to the values that should be used to replace them
+        in the compiled function. The types of the key and value should match or an error will be
+        raised during compilation.
+    """
+    if givens_dict is None:
+        givens_dict = {}
+    # We need a function graph to walk the clients and propagate the volatile property
+    fg = aesara.graph.FunctionGraph(outputs=outputs, clone=False)
+
+    # Walk the graph from inputs to outputs and tag the volatile variables
+    nodes = general_toposort(fg.outputs, deps=lambda x: x.get_parents())
+    givens = []
+    registered_givens = set()
+    volatile_nodes = set()
+    for node in nodes:
+        if (
+            isinstance(node, aesara.compile.SharedVariable)
+            or node in fg.outputs
+            or node in givens_dict
+        ):
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Populate the givens list if the node is in the givens_dict
+        if node in givens_dict and node not in registered_givens:
+            value = givens_dict[node]
+            if not isinstance(value, (Apply, Variable)):
+                # If the value is not a tensor, it is wrapped as a Constant with the node's name
+                value = at.constant(
+                    givens_dict[node],
+                    dtype=node.dtype,
+                    name=node.name,
+                )
+            givens.append((node, value))
+            registered_givens.add(node)
+
+        # Propagate the volatile property to the node's clients
+        if is_volatile:
+            try:
+                clients = [c for c, _ in fg.clients[node]]
+            except KeyError:
+                # Some Apply nodes don't show up in the clients of a FunctionGraph
+                # so we take the Apply node's outputs instead
+                if isinstance(node, aesara.graph.basic.Apply):
+                    clients = node.outputs
+                else:
+                    raise
+            for client in clients:
+                if (
+                    isinstance(client, aesara.graph.basic.Variable)
+                    and client.owner is not None
+                    and isinstance(client.owner.op, at.random.basic.RandomVariable)
+                    and any(rv_param in volatile_nodes for rv_param in client.owner.inputs[3:])
+                ):
```

Comment:
size which is in `inputs[1]`, may be dynamic (i.e., set by a shared variable or another volatile node)

Reply:
Ok, so it should be enough to only look at `inputs[1:]`, right? Because the first input will always be volatile, since it's a shared random stream.

Reply:
I think we might want to just ignore Shared RandomState/Generator Variables altogether. Some RandomVariable-like `ops` created with `OpFromGraph` will include those as the last input instead of the first (e.g., the AR distribution has to do this because of a Scan limitation)

Reply:
How should I approach that then? Filter the inputs that aren't random stream instances?

Reply:
I think so